typing/typecore.ml

```diff
@@ -1770,15 +1770,18 @@ let force_delayed_checks () =
   reset_delayed_checks ();
   Btype.backtrack snap
 
-let rec final_subexpression sexp =
-  match sexp.pexp_desc with
-    Pexp_let (_, _, e)
-  | Pexp_sequence (_, e)
-  | Pexp_try (e, _)
-  | Pexp_ifthenelse (_, e, _)
-  | Pexp_match (_, {pc_rhs=e} :: _)
+let rec final_subexpression exp =
+  match exp.exp_desc with
+    Texp_let (_, _, e)
+  | Texp_sequence (_, e)
+  | Texp_try (e, _)
+  | Texp_ifthenelse (_, e, _)
```

Comment:
The other cases look good, but I don't like the `Texp_ifthenelse` one: it arbitrarily picks the `then` branch to blame, when both the `then` and `else` branches have never-returning types. I'd expect `(if false then assert false else raise Exit); ()` to highlight the whole `if`.

Reply:
The same applies to Text_match, right?  But both were already present in the previous version, this PR should not change  the behavior on those cases.

Reply:
Ah, you're right. I'm still not particularly fond of this behaviour, but this PR does improve it.

Reply:
Thanks!